Draft: add playerctl support

jlo62 opened the pull request:
Hi, this adds support for controlling media.

supports:
* play-pause, play, pause, stop, shuffle, next, prev
* setting the player that should be controlled

todo:
 - [ ] make title configureable, preferable with config file stuff

ErikReider commented:
Needs rebase off of main

jlo62 commented:
@ErikReider This MR should be ready.

The only thing that requires testing is the installing of the config files, I don't want to install it manually on my machine as it interferes with the package manager.

jlo62 commented:
Should also close #116 as it adds a documented example

ErikReider commented on src/mpris-backend/mod.rs:
```diff
@@ -0,0 +1,209 @@
+		})
+	}
+	pub fn run(&mut self) -> Result<(), Box<dyn Error>> {
+		use PlaybackStatus::*;
+		use PlayerctlAction::*;
```
Why the `use` in the function?

ErikReider commented on src/mpris-backend/mod.rs:
```diff
@@ -0,0 +1,209 @@
+		};
+
+		self.icon = Some(icon.to_string());
+		let label = if let Ok(metadata) = metadata {
+			Some(self.fmt_string(metadata))
+		} else {
+			None
+		};
```
The player name or something similar should be displayed instead when running next/prev due to the wrong title being displayed. Ex:

Playing song "A" and running the next command displays the "A" song while it should display "B".

Not really sure if displaying B is possible without delaying and waiting for the next metadata to come in, so an easy fix would be to display the player name :)

ErikReider reviewed (changes requested):
Otherwise, LGTM! :D

ErikReider commented on src/mpris-backend/mod.rs:
```diff
@@ -0,0 +1,237 @@
+                vars.insert("discNumber".to_string(), &disc_num);
+                vars.insert("autoRating".to_string(), &autorating);
+
+                self.fmt_str
+                        .clone()
+                        .unwrap_or("{artist} - {title}".into())
+                        .format(&vars)
+                        .unwrap_or_else(|e| {
+                                eprintln!("error: {}. using default string", e);
+                                "{artist} - {title}".format(&vars).unwrap()
+                        })
+        }
+}
+
+impl PlayerctlAction {
+        pub fn from(action: &str) -> Result<Self, String> {
+                use PlayerctlAction::*;
+                match action {
+                        "play-pause" => Ok(PlayPause),
+                        "play" => Ok(Play),
+                        "pause" => Ok(Pause),
+                        "stop" => Ok(Stop),
+                        "next" => Ok(Next),
+                        "prev" | "previous" => Ok(Prev),
+                        "shuffle" => Ok(Shuffle),
+                        x => Err(x.to_string()),
+                }
+        }
+}
+
+impl PlayerctlDeviceRaw {
+        pub fn from(player: String) -> Result<Self, ()> {
+                use PlayerctlDeviceRaw::*;
+                match player.as_str() {
+                        "auto" | "" => Ok(None),
+                        "all" => Ok(All),
+                        _ => Ok(Some(player)),
+                }
+        }
+}
```
Please use tabs. Run rustfmt on the whole project

ErikReider commented:
Also rebase off of main please

jlo62 commented:
I get this error when rebasing off main:
```
error[E0432]: unresolved import `zbus::ConnectionBuilder`
  --> src/server/main.rs:36:23
   |
36 | use zbus::{interface, ConnectionBuilder};
   |                       ^^^^^^^^^^^^^^^^^ no `ConnectionBuilder` in the root
```

ErikReider commented on src/server/main.rs:
```diff
@@ -59,7 +59,7 @@ impl DbusServer {
 impl DbusServer {
 	async fn new(sender: Sender<(ArgTypes, String)>) -> zbus::Result<()> {
-		let _connection = connection::Builder::session()?
+		let _connection = ConnectionBuilder::session()?
```
You rebased off of main incorrectly, bringing old changes into the mix

ErikReider reviewed (changes requested):
Also, it makes my life easier if you don't force push commits and use proper commit messages :)

jlo62 commented:
@ErikReider I reverted the force push (I think I was rebasing off a ancient version that I hadn't updated in a while) ... I also reverted the changes I didn't do manually. glhf I'm sure I overlooked something.

jlo62 commented:
@ErikReider gentle ping not to forget...

ErikReider reviewed (approved):
LGTM! Thanks for your patience 😅
